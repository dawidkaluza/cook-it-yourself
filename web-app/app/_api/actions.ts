"use server"

import {cookies} from "next/headers";

export function isSignedIn(){
  const cookieStore = cookies();
  const sessionCookie = process.env.NEXT_PUBLIC_API_GATEWAY_SESSION_COOKIE;
  return sessionCookie ? cookieStore.has(sessionCookie) : false;
}