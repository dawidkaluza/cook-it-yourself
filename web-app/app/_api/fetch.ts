import {redirect} from "next/navigation";
import {cookies} from "next/headers";

type ApiRequest = {
  endpoint: string;
  method?: string;
  body?: any;
  headers?: object;
  ignoreAuth?: boolean;
};

type InternalApiRequest = ApiRequest & { baseUrl: string };

export class ApiError extends Error {
  private _response : Response
  constructor(response : Response) {
    super("Client error " + response.status);
    this._response = response;
  }

  get response(): Response {
    return this._response;
  }

  set response(value: Response) {
    this._response = value;
  }
}

function fetchApi({ baseUrl, endpoint, method, body, headers, ignoreAuth } : InternalApiRequest): Promise<any> {
  return new Promise((resolve, reject) => {
    fetch(
      baseUrl + endpoint,
      {
        method: method ?? "GET",
        body: body ? JSON.stringify(body) : undefined,
        headers: {
          "Accept": "application/json",
          "Content-Type": "application/json",
          ...headers,
        },
      }
    ).then(response => {
      if (response.ok) {
        if (!response.body) {
          resolve(null);
        } else {
          response.json()
            .then(data => {
              resolve(data)
            });
        }
      } else {
        switch (response.status) {
          case 401: {
            if (ignoreAuth) {
              reject(new ApiError(response));
            } else {
              redirect("/sign-in");
            }
            break;
          }

          case 403: {
            if (ignoreAuth) {
              reject(new ApiError(response));
            } else {
              redirect("/sign-in");
            }
            break;
          }

          default: {
            reject(new ApiError(response) );
            break;
          }
        }
      }
    }).catch(error => {
      reject(error);
    });
  });
}

export async function fetchFromServer(request : ApiRequest) {
  const { headers } = request;
  const baseUrl = process.env.API_GATEWAY_SERVER_URL;
  if (!baseUrl) {
    throw new Error("API_GATEWAY_SERVER_URL env var not defined");
  }

  return await fetchApi({
    ...request,
    baseUrl,
    headers: {
      ...headers,
      "Cookie": cookies().toString()
    }
  });
}

export async function fetchFromClient(request : ApiRequest) {
  const baseUrl = process.env.NEXT_PUBLIC_API_GATEWAY_CLIENT_URL;
  if (!baseUrl) {
    throw new Error("NEXT_PUBLIC_API_GATEWAY_CLIENT_URL env var not defined");
  }

  return await fetchApi({
    ...request,
    baseUrl,
  });
}
