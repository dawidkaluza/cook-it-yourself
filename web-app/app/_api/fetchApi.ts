function fetchApi(endpoint: string, method?: string, body?: any, headers?: object): Promise<any> {
  return new Promise((resolve, reject) => {
    let baseUrl;

    const isClient = typeof window !== 'undefined';
    if (isClient) {
      baseUrl = process.env.NEXT_PUBLIC_API_GATEWAY_CLIENT_URL;
    } else {
      baseUrl = process.env.API_GATEWAY_SERVER_URL;
    }

    fetch(
      baseUrl + endpoint,
      {
        method: method ?? "GET",
        body: body ? JSON.stringify(body) : undefined,
        headers: {
          "Accept": "application/json",
          "Content-Type": "application/json",
          ...headers,
        },
      }
    ).then(response => {
      if (response.ok) {
        if (!response.body) {
          resolve(null);
        } else {
          response.json()
            .then(data => resolve(JSON.parse(data)));
        }
      } else {
        switch (response.status) {
          case 401: {
            console.error("Unauthorized");
            reject({
              unauthorized: true,
              response,
            });
            break;
          }

          case 403: {
            console.error("Access denied");
            reject({
              accessDenied: true,
              response,
            });
            break;
          }

          default: {
            reject({ response });
            break;
          }
        }
      }
    }).catch(error => {
      reject(error);
    });
  });
}

export { fetchApi };