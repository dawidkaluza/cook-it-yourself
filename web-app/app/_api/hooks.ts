import {useCallback, useEffect, useState} from "react";

export const useAuth = (isAlreadySignedIn?: boolean) => {
  const [isSignedIn, setSignedIn] = useState(!!isAlreadySignedIn);
  const [name, setName] = useState<string | undefined>(undefined);

  useEffect(() => {
    if (!isAlreadySignedIn) {
      const name = sessionStorage.getItem("name");
      setSignedIn(name !== null);
      setName(name ?? undefined);
    }
  }, []);

  const signIn = useCallback((name?: string) => {
    setSignedIn(true);
    setName(name);
    sessionStorage.setItem("name", name ?? "");
  }, []);

  const signOut = useCallback(() => {
    sessionStorage.removeItem("name");
    setSignedIn(false);
    setName(undefined);
  }, []);

  return { isSignedIn, name, signIn, signOut };
};