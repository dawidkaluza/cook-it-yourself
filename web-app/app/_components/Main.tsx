import React from "react";

const Main = ({
  children
}: Readonly<{
  children: React.ReactNode
}>) => {
  return (
    <main style={{margin: "1em"}}>
      {children}
    </main>
  );
};

export { Main };