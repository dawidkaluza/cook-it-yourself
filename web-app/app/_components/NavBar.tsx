"use client"

import React from "react";
import Link from "next/link";
import {usePathname} from "next/navigation";
import {useAuth} from "@/app/_hooks/auth/useAuth";

const NavBar = () => {
  const { isSignedIn } = useAuth();

  return (
    <nav className="navbar navbar-expand bg-body-tertiary">
      <div className="container-fluid text-center">
        <Link href="/" className="navbar-brand">Cook it yourself</Link>
        {isSignedIn ? <NavBarMenu /> : <SignInButton />}
      </div>
    </nav>
  );
};

const NavBarMenu = () => {
  return (
    <>
      <button className="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar-menu"
              aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
        <span className="navbar-toggler-icon"></span>
      </button>

      <div className="collapse navbar-collapse" id="navbar-menu">
        <ul className="navbar-nav">
          <NavBarItem path="/my-recipes">My recipes</NavBarItem>
          <NavBarItem path="/cooking">Cooking</NavBarItem>
        </ul>
      </div>
    </>
  );
};

const NavBarItem = ({
  path,
  children
}: Readonly<{
  path: string,
  children: React.ReactNode,
}>) => {
  const currentPath = usePathname();

  return (
    <li className="nav-item">
      <Link className={path === currentPath ? "nav-link active" : "nav-link"} href={path}>{children}</Link>
    </li>
  );
};

const SignInButton = () => {
  return (
    <a
      href={process.env.API_GATEWAY_CLIENT_URL + "/oauth2/authorization/ciy"}
      className="btn btn-primary btn-sm"
    >
      Sign in
    </a>
  );
};

export {NavBar};