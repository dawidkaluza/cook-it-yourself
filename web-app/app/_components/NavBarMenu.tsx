"use client";

import React from "react";
import {usePathname} from "next/navigation";
import Link from "next/link";

export const NavBarMenu = () => {
  return (
    <>
      <button className="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar-menu"
              aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
        <span className="navbar-toggler-icon"></span>
      </button>

      <div className="collapse navbar-collapse" id="navbar-menu">
        <ul className="navbar-nav">
          <NavBarItem path="/my-recipes">My recipes</NavBarItem>
        </ul>
      </div>
    </>
  );
};

const NavBarItem = ({
  path,
  children
}: Readonly<{
  path: string,
  children: React.ReactNode,
}>) => {
  const currentPath = usePathname();

  return (
    <li className="nav-item">
      <Link className={path === currentPath ? "nav-link active" : "nav-link"} href={path}>{children}</Link>
    </li>
  );
};