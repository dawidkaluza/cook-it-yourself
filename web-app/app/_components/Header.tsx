import React from "react";
import {isSignedIn} from "@/app/_api/auth";
import Link from "next/link";

const Header = () => {
  const signedIn = isSignedIn();

  return (
    <header className="navbar navbar-expand bg-body-tertiary">
      <div className="container-fluid text-center">
        <Link href="/" className="navbar-brand">Cook it yourself</Link>
        {signedIn
          ? <SignOutButton />
          : <SignInButton />
        }
      </div>
    </header>
  );
}

const SignInButton = () => {
  // noinspection HtmlUnknownTarget
  return (
    <a
      href="/sign-in"
      className="btn btn-primary btn-sm"
    >
      Sign in
    </a>
  );
};


const SignOutButton = () => {
  // noinspection HtmlUnknownTarget
  return (
    <a
      href="/sign-out"
      className="btn btn-primary btn-sm"
    >
      Sign out
    </a>
  );
};

export {Header};