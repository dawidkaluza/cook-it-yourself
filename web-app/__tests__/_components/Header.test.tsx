import {Mock} from 'vitest'
import {render, screen} from "@testing-library/react";
import {isSignedIn} from "@/app/_api/auth";
import {Header} from "@/app/_components/Header";

vi.mock("@/app/_api/auth", () => {
  return {
    isSignedIn: vi.fn(),
  };
});

describe('Header component', () => {
  test('render when signed out', () => {
    // Given
    (isSignedIn as Mock).mockImplementation(() => false);

    // When
    const navBar = render(<Header />);

    // Then
    const indexLink = screen.getByRole("link", { name: "Cook it yourself" });
    expect(indexLink).not.toBeNull();
    expect(indexLink?.getAttribute("href")).toBe("/");

    const signInLink = screen.getByRole("link", { name: "Sign in"});
    expect(signInLink).not.toBeNull();
    expect(signInLink?.getAttribute("href")).toBe("/sign-in");

    const signOutLink = screen.queryByRole("link", { name: "Sign out"});
    expect(signOutLink).toBeNull();

    expect(navBar.container).toMatchSnapshot();
  });

  test('render when signed in', () => {
    // Given
    (isSignedIn as Mock).mockImplementation(() => true);

    // When
    const navBar = render(<Header />);

    // Then
    const indexLink = screen.getByRole("link", { name: "Cook it yourself" });
    expect(indexLink).not.toBeNull();
    expect(indexLink?.getAttribute("href")).toBe("/");

    const signInLink = screen.queryByRole("link", { name: "Sign in"});
    expect(signInLink).toBeNull();

    const signOutLink = screen.getByRole("link", { name: "Sign out"});
    expect(signOutLink).not.toBeNull();
    expect(signOutLink?.getAttribute("href")).toBe("/sign-out");

    expect(navBar.container).toMatchSnapshot();
  });
});