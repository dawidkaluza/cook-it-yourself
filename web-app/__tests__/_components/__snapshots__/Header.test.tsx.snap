// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header component > render when signed in 1`] = `
<div>
  <header
    class="navbar navbar-expand bg-body-tertiary"
  >
    <div
      class="container-fluid text-center"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Cook it yourself
      </a>
      <a
        class="btn btn-primary btn-sm"
        href="/sign-out"
      >
        Sign out
      </a>
    </div>
  </header>
</div>
`;

exports[`Header component > render when signed out 1`] = `
<div>
  <header
    class="navbar navbar-expand bg-body-tertiary"
  >
    <div
      class="container-fluid text-center"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Cook it yourself
      </a>
      <a
        class="btn btn-primary btn-sm"
        href="/sign-in"
      >
        Sign in
      </a>
    </div>
  </header>
</div>
`;
